colony/utils.py

```diff
@@ -150,3 +182,96 @@ def parse_comma_separated_string(params_string: str = None) -> dict:
         res[key] = val
 
     return res
+
+
+def figure_out_branch(branch, name):
+    temp_working_branch = ""
+    repo = None
+    if branch:
+        working_branch = branch
+    else:
+        # Try to detect branch from current git-enabled folder
+        logger.debug("Branch hasn't been specified. Trying to identify branch from current working directory")
+        try:
+            repo = BlueprintRepo(os.getcwd())
+            # todo get flag for stash mode
+            working_branch = get_blueprint_working_branch(repo, blueprint_name=name)
+            BaseCommand.message(f"Automatically detected current working branch: {working_branch}")
+
+        except BadBlueprintRepo as e:
+            working_branch = None
+            logger.warning(
+                f"No branch has been specified and it could not be identified from the working directory; "
+                f"reason: {e}. A branch of the Blueprints Repository attached to Colony Space will be used"
+            )
+
+        # Checking if:
+        # 1) User has specified not use local (specified a branch)
+        # 2) User is in an actual git dir (working_branch)
+        # 3) There is even a need to create a temp branch for out-of-sync reasons:
+        #   either repo.is_dirty() (changes have not been committed locally)
+        #   or not repo.is_current_branch_synced() (changes committed locally but not pushed to remote)
+        if not branch and working_branch and (repo.is_dirty() or not repo.is_current_branch_synced()):
```

Comment:
Unfortunately this condition still has a couple of things we need to address:

1. If a user is working in a local branch which doesn't exist in remote , this block will be unreachable (`working_branch is None`). However, it's a valid case for _temporary_branch_ feature

2. It turned out that is_dirty() checks only changes in already existing and tracked files. But if user has added some new files but hasn't tracked them (`git add file`) -> `is_dirty()` will be `false`. I think it's a valid case for temp branch creation as well.
So, in my opinion the reasons why we want to have a temp branch are the following:
  - you are working in the branch which exists on remote but branches are not synced. **covered**
  - you are working in a branch which doesn't exist on remote
  - repo is dirty. There are new tracked files or changes in existing files. **covered**
  - repo has untracked files. there is a property to check it `repo.untracked_files`